Add sensuctl prune to sensuctl reference doc (#2305)
sensu/sensu-docs: content/sensu-go/5.19/sensuctl/reference.md

- [Export resources](#export-resources)
- [Manage resources](#manage-resources)
  - [Subcommands](#subcommands)
- [Prune resources](#prune-resources) (experimental feature)

> - [Response filtering](#response-filtering)

Reviewer: I'm not sure if you meant to remove `(commercial feature)` from response filtering here?

Author: Yeah, I should have removed it sooner, but I should probably remove it from all versions in a different PR. I'll put it back here. We have the commercial features note in the response filtering text. Thank you for checking.

See the [RBAC reference][22] for information about local user management with sensuctl.

#### sensuctl prune

{{% notice important %}}
**IMPORTANT**: `sensuctl prune` is an alpha feature in release 5.19.0 and may include breaking changes.
{{% /notice %}}

**COMMERCIAL FEATURE**: Access sensuctl pruning in the packaged Sensu Go distribution.
For more information, see [Get started with commercial features][30].

The `sensuctl prune` subcommand allows you to delete resources that do not appear in a given set of Sensu objects (called a "configuration") from a from a file, URL, or STDIN.
For example, you can use `sensuctl create` to to apply a new configuration, then use `sensuctl prune` to prune unneeded resources, resources that were created by a specific user or that include a specific label selector, and more.

`sensuctl prune` can only delete resources that have the label `sensu.io/managed_by: sensuctl`, which Sensu automatically adds to all resources created with sensuctl.

> This means you can only use `sensuctl prune` for delete resources that were created with sensuctl.

Reviewer: You would know better than me in terms of grammar, is `for` or `to` the correct usage here?
```suggestion
This means you can only use `sensuctl prune` to delete resources that were created with sensuctl.
```

Author: no you are correct! I probably changed "deleting" to "delete" and forgot to fix the "for" 
Thank you for checking

The pruning operation always follows the role-based access control (RBAC) permissions of the current user.
For example, to prune resources in the `dev` namespace, the current user who sends the prune command must have delete access to the `dev` namespace.

##### sensuctl prune usage

{{< highlight shell >}}

> sensuctl prune [RESOURCE TYPE],[RESOURCE TYPE]... [-f [FILE] -f [URL] -f [STDIN] [-r] ... ] [--NAMESPACE] [flags]

Author: ```suggestion
sensuctl prune [RESOURCE TYPE],[RESOURCE TYPE]... -f [FILE or URL] [-r] ... ] [--NAMESPACE] [flags]
```

> sensuctl prune [RESOURCE TYPE],[RESOURCE TYPE]... [-f [FILE] -f [URL] -f [STDIN] [-r] ... ] [--NAMESPACE] [flags]

Reviewer: `-f` is not required for data from STDIN, rather data would likely be piped to the command via standard in (ex. `echo check.yml | sensuctl prune`). You probably don't need an example of STDIN usage here, but we should at the very least remove `-f [STDIN]`.

{{< /highlight >}}

In this example `sensuctl prune` command:

- Replace [RESOURCE TYPE] with the [synonym or fully qualified name][48] of the resource you want to prune.
You must specify at least one resource type or the `all` qualifier (to prune all resource types).

> - Replace [FILE], [URL], or [STDIN] with the name of the file, URL, or STDIN where you want to apply the pruning.

Author: ```suggestion
- Replace [FILE or URL] with the name of the file or the URL where you want to apply the pruning.
```

> - Replace [FILE], [URL], or [STDIN] with the name of the file, URL, or STDIN where you want to apply the pruning.

Reviewer: https://github.com/sensu/sensu-docs/pull/2305/files#r398154539

- Replace [flags] with the flags you want to use, if any.

> - Replace [--NAMESPACE] with the namespace where you want to apply pruning.

Reviewer: This defaults to the current configured namespace if the flag is omitted.

> To prune all namespaces, omit the namespace qualifier from your command.

Author: ```suggestion
If you omit the namespace qualifier, the command defaults to the current configured namespace.
```

> To prune all namespaces, omit the namespace qualifier from your command.

Reviewer: Sensuctl does not currently support pruning from multiple namespaces. Users should issue multiple `sensuctl prune` commands in each namespace they want to prune resources from. There is work to enhance this with a `--all-namespaces` and/or `--namespaces` flag in the future.

Use a comma separator to prune more than one resource in a single command.

For example, to prune checks and assets from the file `checks.yaml` for the `dev` namespace and the `admin` and `ops` users:

{{< highlight shell >}}
sensuctl prune checks,assets --file checks.yaml --namespace dev --users admin,ops
{{< /highlight >}}

##### sensuctl prune flags

Run `sensuctl prune -h` to view command-specific and global flags.
The following table describes the command-specific flags.

> | Flag | Function and important notes

Reviewer: Snuck in another flag after this issue was filed, sorry! Cluster wide defaults to false, but if set it prunes any cluster wide (non-namespaced) resources that are not defined in the configuration.
```
-c, --cluster-wide            Prunes cluster-wide resources
```

| ---- | ----------------------------
`-a` or `--all-users` | Prunes resources created by all users. Mutually exclusive with the `--users` flag. Defaults to false.

> `-d` or `--dry-run` | Prints the resources that will be pruned but does not actually delete them. Defaults to false.

Author: ```suggestion
`-c` or `--cluster-wide` | Prunes any cluster-wide (non-namespaced) resources that are not defined in the configuration. Defaults to false.
`-d` or `--dry-run` | Prints the resources that will be pruned but does not actually delete them. Defaults to false.
```

See the [RBAC reference][22] for information about local user management with sensuctl.

## Prune resources

{{% notice important %}}
**IMPORTANT**: `sensuctl prune` is experimental and under testing for Sensu Go 5.19.0.

> We may remove this feature from 5.19.0 and future releases.

Reviewer: I don't think we anticipate removing the feature (although I'm sure we could). I would try rephrasing this to frame that `sensuctl prune` is an alpha feature where breaking changes are still permitted.

{{% /notice %}}

**COMMERCIAL FEATURE**: Access `sensuctl prune` in the packaged Sensu Go distribution.
For more information, see [Get started with commercial features][30].

The `sensuctl prune` command allows you to delete resources that do not appear in your configuration from a file, URL, or STDIN.

> For example, you can use `sensuctl prune` to prune resources that were created by a specific user or that include a specific label selector.

Reviewer: One point I forgot to mention:
- `sensuctl prune` will only prune resources that were created with `sensuctl create` or resources that have the following label: 
```
labels:
    sensu.io/managed_by: sensuctl

"labels": {
      "sensu.io/managed_by": "sensuctl"
    },
```

### sensuctl prune usage

{{< highlight shell >}}
sensuctl prune [RESOURCE TYPE],[RESOURCE TYPE]... [-f [FILE] -f [URL] -f [URL] [-r] ... ] [flags]
{{< /highlight >}}

In this example `sensuctl prune` command:

- Replace [RESOURCE TYPE] with the [synonym or fully qualified name][48] of the resource you want to prune.

> If you do not specify any resource types, `sensuctl prune` defaults to all types except events and entities.

Reviewer: If no resource types are provided, the command will actually error. In order to prune all resource types, the `all` keyword must be used.

- Replace [FILE], [URL], or [STDIN] with the name of the file, URL, or STDIN where you want to apply the pruning.
- Replace [flags] with the flags you want to use, if any.

Use a comma separator to prune more than one resource in a single command.

For example, to prune checks and assets from the file `checks.yaml` for the `dev` namespace and the `admin` and `ops` users:

{{< highlight shell >}}
sensuctl prune checks,assets --file checks.yaml --namespace dev --users admin,ops
{{< /highlight >}}

### sensuctl prune flags

Run `sensuctl prune -h` to view command-specific and global flags.

> The `sensuctl prune` flags have a few important characteristics:

Reviewer: I think this might be better suited as a table, so then we can define each flag, if its required, what the default is, and what the behavior is?

> - `--all-users` defaults to false and overrides the `--users` flag.

Reviewer: ```suggestion
- `--all-users` defaults to false and is mutually exclusive with the `--users` flag.
```
